### PR TITLE
HADOOP-19014. Jackson 2.14.3 requiring jsr311-compat

### PR DESCRIPTION
**Pull request by pjfanning**
### Description of PR

HADOOP-19014

see https://github.com/pjfanning/jsr311-compat/

### For code changes:

- [x] Does the title or this PR starts with the corresponding JIRA issue id (e.g. 'HADOOP-17799. Your PR title ...')?
- [ ] Object storage: have the integration tests been executed and the endpoint declared according to the connector-specific documentation?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] If applicable, have you updated the `LICENSE`, `LICENSE-binary`, `NOTICE-binary` files?

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |   0m 33s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  0s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  0s |  |  detect-secrets was not available.  |
| +0 :ok: |  xmllint  |   0m  0s |  |  xmllint was not available.  |
| +0 :ok: |  shelldocs  |   0m  0s |  |  Shelldocs was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  14m 43s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  30m 44s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  16m 22s |  |  trunk passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04  |
| +1 :green_heart: |  compile  |  14m 51s |  |  trunk passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  mvnsite  |  18m 50s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   8m 28s |  |  trunk passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04  |
| +1 :green_heart: |  javadoc  |   7m 35s |  |  trunk passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  shadedclient  |  47m 32s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 48s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |  43m 51s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  15m 33s |  |  the patch passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04  |
| +1 :green_heart: |  javac  |  15m 33s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  14m 55s |  |  the patch passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  javac  |  14m 55s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  mvnsite  |  14m 27s |  |  the patch passed  |
| +1 :green_heart: |  shellcheck  |   0m  0s |  |  No new issues.  |
| +1 :green_heart: |  javadoc  |   8m 16s |  |  the patch passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04  |
| +1 :green_heart: |  javadoc  |   7m 32s |  |  the patch passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  shadedclient  |  48m 52s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  | 738m 33s | [/patch-unit-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6370/1/artifact/out/patch-unit-root.txt) |  root in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m 36s |  |  The patch does not generate ASF License warnings.  |
|  |   | 1027m 14s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.yarn.server.resourcemanager.reservation.TestCapacityOverTimePolicy |
|   | hadoop.hdfs.server.balancer.TestBalancerWithHANameNodes |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6370/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/6370 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient codespell detsecrets xmllint shellcheck shelldocs |
| uname | Linux 359bc3484ca3 5.15.0-86-generic #96-Ubuntu SMP Wed Sep 20 08:23:49 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / 025ad17c8ed282fa9e69dd292e2043df5ceddf90 |
| Default Java | Private Build-1.8.0_392-8u392-ga-1~20.04-b08 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_392-8u392-ga-1~20.04-b08 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6370/1/testReport/ |
| Max. process+thread count | 3719 (vs. ulimit of 5500) |
| modules | C: hadoop-project hadoop-common-project/hadoop-common hadoop-yarn-project/hadoop-yarn/hadoop-yarn-common hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-applicationhistoryservice hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager hadoop-yarn-project/hadoop-yarn/hadoop-yarn-applications/hadoop-yarn-applications-catalog/hadoop-yarn-applications-catalog-webapp hadoop-client-modules/hadoop-client hadoop-tools/hadoop-resourceestimator hadoop-client-modules/hadoop-client-minicluster . U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6370/1/console |
| versions | git=2.25.1 maven=3.6.3 shellcheck=0.7.0 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

File: hadoop-project/pom.xml [hunk 1]
```diff
@@ -921,6 +921,11 @@
           </exclusion>
         </exclusions>
       </dependency>
+      <dependency>
```
Comment (steveloughran) [inline hadoop-project/pom.xml:924]: should have a comment above to explain, maybe something on L72 to say why we've had problems with jackson updates

File: hadoop-common-project/hadoop-common/pom.xml [hunk 2]
```diff
@@ -179,6 +179,11 @@
         </exclusion>
       </exclusions>
     </dependency>
+    <dependency>
+      <groupId>com.github.pjfanning</groupId>
+      <artifactId>jsr311-compat</artifactId>
+      <scope>runtime</scope>
```
Comment (steveloughran) [inline hadoop-common-project/hadoop-common/pom.xml:185]: what will break if this is not on the classpath?

**Comment by steveloughran** [comment]
I'm happy with this, just need to understand if there are any problems with it being optional

we will need to add in release notes as incompatible and explain the issue and why the jar should be excluded if you do have the jax-rs stuff on the cp, and that findclass can always be used to work out where the class is from

File: hadoop-project/pom.xml (hunk 1 above)
Comment (pjfanning) [reply hadoop-project/pom.xml:924]: I'll add some comments and recheck the dependency scopes.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime |  Logfile | Comment |
|:----:|----------:|--------:|:--------:|:-------:|
| +0 :ok: |  reexec  |  11m 19s |  |  Docker mode activated.  |
|||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  |  No case conflicting files found.  |
| +0 :ok: |  codespell  |   0m  1s |  |  codespell was not available.  |
| +0 :ok: |  detsecrets  |   0m  1s |  |  detect-secrets was not available.  |
| +0 :ok: |  xmllint  |   0m  1s |  |  xmllint was not available.  |
| +0 :ok: |  shelldocs  |   0m  1s |  |  Shelldocs was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
|||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |  14m 25s |  |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  31m 24s |  |  trunk passed  |
| +1 :green_heart: |  compile  |  16m 20s |  |  trunk passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04  |
| +1 :green_heart: |  compile  |  14m 52s |  |  trunk passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  mvnsite  |  18m 21s |  |  trunk passed  |
| +1 :green_heart: |  javadoc  |   8m 26s |  |  trunk passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04  |
| +1 :green_heart: |  javadoc  |   7m 32s |  |  trunk passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  shadedclient  |  47m 23s |  |  branch has no errors when building and testing our client artifacts.  |
|||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 49s |  |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |  44m  5s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  15m 45s |  |  the patch passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04  |
| +1 :green_heart: |  javac  |  15m 45s |  |  the patch passed  |
| +1 :green_heart: |  compile  |  14m 59s |  |  the patch passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  javac  |  14m 59s |  |  the patch passed  |
| +1 :green_heart: |  blanks  |   0m  0s |  |  The patch has no blanks issues.  |
| +1 :green_heart: |  mvnsite  |  13m 55s |  |  the patch passed  |
| +1 :green_heart: |  shellcheck  |   0m  0s |  |  No new issues.  |
| +1 :green_heart: |  javadoc  |   8m 17s |  |  the patch passed with JDK Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04  |
| +1 :green_heart: |  javadoc  |   7m 30s |  |  the patch passed with JDK Private Build-1.8.0_392-8u392-ga-1~20.04-b08  |
| +1 :green_heart: |  shadedclient  |  48m 59s |  |  patch has no errors when building and testing our client artifacts.  |
|||| _ Other Tests _ |
| -1 :x: |  unit  | 747m 25s | [/patch-unit-root.txt](https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6370/2/artifact/out/patch-unit-root.txt) |  root in the patch passed.  |
| +1 :green_heart: |  asflicense  |   1m 37s |  |  The patch does not generate ASF License warnings.  |
|  |   | 1046m 59s |  |  |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.yarn.server.nodemanager.webapp.TestNMWebServicesApps |
|   | hadoop.yarn.server.nodemanager.webapp.TestNMWebServicesContainers |
|   | hadoop.yarn.server.nodemanager.webapp.TestNMWebServicesAuxServices |
|   | hadoop.yarn.server.nodemanager.webapp.TestNMWebServices |
|   | hadoop.yarn.server.applicationhistoryservice.webapp.TestAHSWebServices |
|   | hadoop.yarn.server.resourcemanager.webapp.TestRMWebServicesCapacitySchedDynamicConfig |
|   | hadoop.yarn.server.resourcemanager.webapp.TestRMWebServicesCapacitySchedLegacyQueueCreation |
|   | hadoop.yarn.server.resourcemanager.webapp.TestRMWebServicesSchedulerActivities |
|   | hadoop.yarn.server.resourcemanager.webapp.TestRMWebServicesAppCustomResourceTypes |
|   | hadoop.yarn.server.resourcemanager.webapp.TestRMWebServices |
|   | hadoop.yarn.server.resourcemanager.webapp.fairscheduler.TestRMWebServicesFairScheduler |
|   | hadoop.yarn.server.resourcemanager.webapp.TestRMWebServicesNodeLabels |
|   | hadoop.yarn.server.resourcemanager.webapp.TestRMWebServicesAppsModification |
|   | hadoop.yarn.server.resourcemanager.webapp.TestRMWebServicesSchedulerActivitiesWithMultiNodesEnabled |
|   | hadoop.yarn.server.resourcemanager.webapp.TestRMWebServicesConfigurationMutation |
|   | hadoop.yarn.server.resourcemanager.webapp.TestRMWebServicesHttpStaticUserPermissions |
|   | hadoop.yarn.server.resourcemanager.TestRMHA |
|   | hadoop.yarn.server.resourcemanager.webapp.TestRMWebServicesApps |
|   | hadoop.yarn.server.resourcemanager.webapp.TestRMWebServicesCapacitySchedulerMixedMode |
|   | hadoop.yarn.server.resourcemanager.webapp.TestRMWebServiceAppsNodelabel |
|   | hadoop.yarn.server.resourcemanager.webapp.TestRMWebServicesForCSWithPartitions |
|   | hadoop.yarn.server.resourcemanager.webapp.TestRMWebServicesCapacitySched |
|   | hadoop.yarn.server.resourcemanager.webapp.TestRMWebServicesAppsCustomResourceTypes |
|   | hadoop.yarn.server.resourcemanager.webapp.fairscheduler.TestRMWebServicesFairSchedulerCustomResourceTypes |
|   | hadoop.yarn.server.resourcemanager.webapp.TestRMWebServicesNodes |
|   | hadoop.yarn.server.resourcemanager.webapp.TestRMWebServicesDelegationTokens |
|   | hadoop.yarn.server.resourcemanager.webapp.TestRMWebServicesAppAttempts |
|   | hadoop.yarn.server.globalpolicygenerator.webapp.TestGPGWebServices |
|   | hadoop.mapreduce.v2.app.webapp.TestAMWebServices |
|   | hadoop.mapreduce.v2.app.webapp.TestAMWebServicesAttempt |
|   | hadoop.mapreduce.v2.app.webapp.TestAMWebServicesJobs |
|   | hadoop.mapreduce.v2.app.webapp.TestAMWebServicesAttempts |
|   | hadoop.mapreduce.v2.app.webapp.TestAMWebServicesJobConf |
|   | hadoop.mapreduce.v2.app.webapp.TestAMWebServicesTasks |
|   | hadoop.mapreduce.v2.hs.webapp.TestHsWebServicesJobsQuery |
|   | hadoop.mapreduce.v2.hs.webapp.TestHsWebServicesLogs |
|   | hadoop.mapreduce.v2.hs.webapp.TestHsWebServicesJobConf |
|   | hadoop.mapreduce.v2.hs.webapp.TestHsWebServicesTasks |
|   | hadoop.mapreduce.v2.hs.webapp.TestHsWebServices |
|   | hadoop.mapreduce.v2.hs.webapp.TestHsWebServicesAttempts |
|   | hadoop.mapreduce.v2.hs.webapp.TestHsWebServicesJobs |
|   | hadoop.hdfs.server.datanode.TestDirectoryScanner |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.43 ServerAPI=1.43 base: https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6370/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/6370 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient codespell detsecrets xmllint shellcheck shelldocs |
| uname | Linux 7481ec27e9d7 5.15.0-86-generic #96-Ubuntu SMP Wed Sep 20 08:23:49 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | dev-support/bin/hadoop.sh |
| git revision | trunk / a08b99e8649b8b39c122587fbe3fa70c309d1817 |
| Default Java | Private Build-1.8.0_392-8u392-ga-1~20.04-b08 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.21+9-post-Ubuntu-0ubuntu120.04 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_392-8u392-ga-1~20.04-b08 |
|  Test Results | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6370/2/testReport/ |
| Max. process+thread count | 4218 (vs. ulimit of 5500) |
| modules | C: hadoop-project hadoop-common-project/hadoop-common hadoop-yarn-project/hadoop-yarn/hadoop-yarn-common hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-applicationhistoryservice hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-resourcemanager hadoop-yarn-project/hadoop-yarn/hadoop-yarn-server/hadoop-yarn-server-nodemanager hadoop-yarn-project/hadoop-yarn/hadoop-yarn-applications/hadoop-yarn-applications-catalog/hadoop-yarn-applications-catalog-webapp hadoop-client-modules/hadoop-client hadoop-tools/hadoop-resourceestimator hadoop-client-modules/hadoop-client-minicluster . U: . |
| Console output | https://ci-hadoop.apache.org/job/hadoop-multibranch/job/PR-6370/2/console |
| versions | git=2.25.1 maven=3.6.3 shellcheck=0.7.0 |
| Powered by | Apache Yetus 0.14.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by pjfanning** [comment]
I'm going to come back to this but in a few weeks. There are some incompatibilities showing up in testing and I want to have a closer look. I have a few days off this week so I don't want to delay the 3.4.0 branching.

**Pull request closed without merging**